### ISSUE TITLE
deploy isn't working

### ISSUE THREAD

**Issue by Stopi** [domovikapp/domovik-server#4]
Hi,

I'm following instructions to host my own instance.

The [step 6](https://github.com/domovikapp/domovik-server#build-domovik) (`mix setup`) gave me errors.
It seems this step depends on python-2.
Please add this to the list of requirements or (better) upgrade your script to python-3.

I've installed python 2.7 but I got errors nonetheless.
Please have a look at the npm log produced: [2022-10-20T05_34_53_033Z-debug-0.log](https://github.com/domovikapp/domovik-server/files/9826728/2022-10-20T05_34_53_033Z-debug-0.log).

Or is this a dead project?

**Comment by delehef** [comment]
Hi, and sorry for the delay.

The dependencies issue has been fixed and the requirement for Python 2 made explicit, the manual deployment process should work correctly now.

> or (better) upgrade your script to python-3.

An update to the whole assets generation pipeline and an upgrade to a newer version of Phoenix is under work.

> Or is this a dead project?

No, there are still quite a lot of things I want to add; but I only have 24hrs per day :(

**Comment by Stopi** [comment]
Hi, thanks for your answer.
> No, there are still quite a lot of things I want to add; but I only have 24hrs per day :(

I'm glad to read this.
I mean: glad the project is still alive.
We all are so busy, that's why I was happy to find this project instead of having to code mine.

It worked
--
Today I had some time to try again.
This time the deploy worked.

_Note:_ before runing `mix phx.gen.secret`, one should download the dependecies with `mix deps.get`.

Pow configuration
--
Next step is to configure Pow.
So far instructions are telling me to do things that are already done, unless I need a shared folder for mnesia's database files.

I found a typo, in `config/prod.exs`, on line 53.
`config :domovik, :pow, cache_store_backed: Pow.Store.Backend.MnesiaCache`
should be
`config :domovik, :pow, cache_store_backend: Pow.Store.Backend.MnesiaCache`

Running server
--
Then I launched the server, but it complained about some missing certificate.
```
$ mix phx.server
[error] `inotify-tools` is needed to run `file_system` for your system, check https://github.com/rvoicilas/inotify-tools/wiki for more information about how to install it. If it's already installed but not be found, appoint executable file with `config.exs` or `FILESYSTEM_FSINOTIFY_EXECUTABLE_FILE` env.
[warning] Could not start Phoenix live-reload because we cannot listen to the file system.
You don't need to worry! This is an optional feature used during development to
refresh your browser when you save files and it does not affect production.

[notice] Application domovik exited: Domovik.Application.start(:normal, []) returned an error: shutdown: failed to start child: DomovikWeb.Endpoint
    ** (EXIT) an exception was raised:
        ** (ArgumentError) could not start Cowboy2 adapter, the file /opt/domovik/domovik-server/_build/dev/lib/domovik/priv/cert/selfsigned_key.pem required by SSL's :keyfile either does not exist, or the application does not have permission to access it
            (plug_cowboy 2.5.2) lib/plug/cowboy.ex:386: Plug.Cowboy.fail/1
            (plug_cowboy 2.5.2) lib/plug/cowboy.ex:144: Plug.Cowboy.args/4
            (plug_cowboy 2.5.2) lib/plug/cowboy.ex:238: Plug.Cowboy.child_spec/1
            (phoenix 1.6.5) lib/phoenix/endpoint/cowboy2_adapter.ex:85: Phoenix.Endpoint.Cowboy2Adapter.child_spec/3
            (phoenix 1.6.5) lib/phoenix/endpoint/cowboy2_adapter.ex:65: anonymous fn/5 in Phoenix.Endpoint.Cowboy2Adapter.child_specs/2
            (elixir 1.13.3) lib/enum.ex:2396: Enum."-reduce/3-lists^foldl/2-0-"/3
            (phoenix 1.6.5) lib/phoenix/endpoint/cowboy2_adapter.ex:56: Phoenix.Endpoint.Cowboy2Adapter.child_specs/2
            (phoenix 1.6.5) lib/phoenix/endpoint/supervisor.ex:106: Phoenix.Endpoint.Supervisor.init/1
            (stdlib 4.0.1) supervisor.erl:330: :supervisor.init/1
            (stdlib 4.0.1) gen_server.erl:848: :gen_server.init_it/2
            (stdlib 4.0.1) gen_server.erl:811: :gen_server.init_it/6
            (stdlib 4.0.1) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
[notice] Application stripity_stripe exited: :stopped
[notice] Application uri_query exited: :stopped
[notice] Application bamboo_phoenix exited: :stopped
[notice] Application bamboo exited: :stopped
[notice] Application timex exited: :stopped
[notice] Application combine exited: :stopped
[notice] Application tzdata exited: :stopped
[notice] Application hackney exited: :stopped
[notice] Application metrics exited: :stopped
[notice] Application ssl_verify_fun exited: :stopped
[notice] Application parse_trans exited: :stopped
[notice] Application syntax_tools exited: :stopped
[notice] Application certifi exited: :stopped
[notice] Application mimerl exited: :stopped
[notice] Application idna exited: :stopped
[notice] Application unicode_util_compat exited: :stopped
[notice] Application elixir_uuid exited: :stopped
[notice] Application pow exited: :stopped
[notice] Application plug_cowboy exited: :stopped
[notice] Application cowboy_telemetry exited: :stopped
[notice] Application cowboy exited: :stopped
[notice] Application ranch exited: :stopped
[notice] Application cowlib exited: :stopped
[notice] Application jason exited: :stopped
[notice] Application telemetry_poller exited: :stopped
[notice] Application phoenix_live_dashboard exited: :stopped
[notice] Application telemetry_metrics exited: :stopped
[notice] Application phoenix_live_view exited: :stopped
[notice] Application phoenix_live_reload exited: :stopped
[notice] Application file_system exited: :stopped
[notice] Application phoenix_html exited: :stopped
[notice] Application ecto_psql_extras exited: :stopped
[notice] Application table_rex exited: :stopped
[notice] Application postgrex exited: :stopped
[notice] Application ecto_sql exited: :stopped
[notice] Application db_connection exited: :stopped
[notice] Application connection exited: :stopped
[notice] Application phoenix_ecto exited: :stopped
[notice] Application ecto exited: :stopped
[notice] Application decimal exited: :stopped
[notice] Application mnesia exited: :stopped
[os_mon] memory supervisor port (memsup): Erlang has closed
[os_mon] cpu supervisor port (cpu_sup): Erlang has closed
[notice] Application os_mon exited: :stopped
[notice] Application sasl exited: :stopped
[notice] Application runtime_tools exited: :stopped
** (Mix) Could not start application domovik: Domovik.Application.start(:normal, []) returned an error: shutdown: failed to start child: DomovikWeb.Endpoint
    ** (EXIT) an exception was raised:
        ** (ArgumentError) could not start Cowboy2 adapter, the file /opt/domovik/domovik-server/_build/dev/lib/domovik/priv/cert/selfsigned_key.pem required by SSL's :keyfile either does not exist, or the application does not have permission to access it
            (plug_cowboy 2.5.2) lib/plug/cowboy.ex:386: Plug.Cowboy.fail/1
            (plug_cowboy 2.5.2) lib/plug/cowboy.ex:144: Plug.Cowboy.args/4
            (plug_cowboy 2.5.2) lib/plug/cowboy.ex:238: Plug.Cowboy.child_spec/1
            (phoenix 1.6.5) lib/phoenix/endpoint/cowboy2_adapter.ex:85: Phoenix.Endpoint.Cowboy2Adapter.child_spec/3
            (phoenix 1.6.5) lib/phoenix/endpoint/cowboy2_adapter.ex:65: anonymous fn/5 in Phoenix.Endpoint.Cowboy2Adapter.child_specs/2
            (elixir 1.13.3) lib/enum.ex:2396: Enum."-reduce/3-lists^foldl/2-0-"/3
            (phoenix 1.6.5) lib/phoenix/endpoint/cowboy2_adapter.ex:56: Phoenix.Endpoint.Cowboy2Adapter.child_specs/2
            (phoenix 1.6.5) lib/phoenix/endpoint/supervisor.ex:106: Phoenix.Endpoint.Supervisor.init/1
            (stdlib 4.0.1) supervisor.erl:330: :supervisor.init/1
            (stdlib 4.0.1) gen_server.erl:848: :gen_server.init_it/2
            (stdlib 4.0.1) gen_server.erl:811: :gen_server.init_it/6
            (stdlib 4.0.1) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
```
so I've created a self signed certificate with the required filename:
```
mkdir _build/dev/lib/domovik/priv/cert
cd _build/dev/lib/domovik/priv/cert
openssl req -x509 -nodes -newkey rsa:4096 -keyout selfsigned_key.pem -out selfsigned.pem -sha256 -days 36500
```
Let's run again:
```
$ cd -
$ mix phx.server
[error] `inotify-tools` is needed to run `file_system` for your system, check https://github.com/rvoicilas/inotify-tools/wiki for more information about how to install it. If it's already installed but not be found, appoint executable 
file with `config.exs` or `FILESYSTEM_FSINOTIFY_EXECUTABLE_FILE` env.                                                
[warning] Could not start Phoenix live-reload because we cannot listen to the file system.
You don't need to worry! This is an optional feature used during development to                                      
refresh your browser when you save files and it does not affect production.             
                                                                                                                     
[info] Running DomovikWeb.Endpoint with cowboy 2.9.0 at 0.0.0.0:4000 (http)
[info] Running DomovikWeb.Endpoint with cowboy 2.9.0 at 0.0.0.0:4001 (https)
[info] Access DomovikWeb.Endpoint at https://localhost:4001
[info] Mnesia cluster initiated on :nonode@nohost
[webpack-cli] Error: Unknown option '--watch-stdin'
[webpack-cli] Run 'webpack --help' to see available commands and options
[error] Task #PID<0.797.0> started from DomovikWeb.Endpoint terminating
** (stop) :watcher_command_error
    (phoenix 1.6.5) lib/phoenix/endpoint/watcher.ex:55: Phoenix.Endpoint.Watcher.watch/2
    (elixir 1.13.3) lib/task/supervised.ex:89: Task.Supervised.invoke_mfa/2
    (stdlib 4.0.1) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
Function: &Phoenix.Endpoint.Watcher.watch/2
    Args: ["node", ["node_modules/webpack/bin/webpack.js", "--mode", "development", "--watch-stdin", {:cd, "/opt/domovik/domovik-server/assets"}]]
[webpack-cli] Error: Unknown option '--watch-stdin'
[webpack-cli] Run 'webpack --help' to see available commands and options
[error] Task #PID<0.802.0> started from DomovikWeb.Endpoint terminating
** (stop) :watcher_command_error
    (phoenix 1.6.5) lib/phoenix/endpoint/watcher.ex:55: Phoenix.Endpoint.Watcher.watch/2
    (elixir 1.13.3) lib/task/supervised.ex:89: Task.Supervised.invoke_mfa/2
    (stdlib 4.0.1) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
Function: &Phoenix.Endpoint.Watcher.watch/2
    Args: ["node", ["node_modules/webpack/bin/webpack.js", "--mode", "development", "--watch-stdin", {:cd, "/opt/domovik/domovik-server/assets"}]]
[webpack-cli] Error: Unknown option '--watch-stdin'
[webpack-cli] Run 'webpack --help' to see available commands and options
[error] Task #PID<0.803.0> started from DomovikWeb.Endpoint terminating
** (stop) :watcher_command_error
    (phoenix 1.6.5) lib/phoenix/endpoint/watcher.ex:55: Phoenix.Endpoint.Watcher.watch/2
    (elixir 1.13.3) lib/task/supervised.ex:89: Task.Supervised.invoke_mfa/2
    (stdlib 4.0.1) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
Function: &Phoenix.Endpoint.Watcher.watch/2
    Args: ["node", ["node_modules/webpack/bin/webpack.js", "--mode", "development", "--watch-stdin", {:cd, "/opt/domovik/domovik-server/assets"}]]
[webpack-cli] Error: Unknown option '--watch-stdin'
[webpack-cli] Run 'webpack --help' to see available commands and options
[error] Task #PID<0.804.0> started from DomovikWeb.Endpoint terminating
** (stop) :watcher_command_error
    (phoenix 1.6.5) lib/phoenix/endpoint/watcher.ex:55: Phoenix.Endpoint.Watcher.watch/2
    (elixir 1.13.3) lib/task/supervised.ex:89: Task.Supervised.invoke_mfa/2
    (stdlib 4.0.1) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
Function: &Phoenix.Endpoint.Watcher.watch/2
    Args: ["node", ["node_modules/webpack/bin/webpack.js", "--mode", "development", "--watch-stdin", {:cd, "/opt/domovik/domovik-server/assets"}]]
[webpack-cli] Error: Unknown option '--watch-stdin'
[webpack-cli] Run 'webpack --help' to see available commands and options

```
It's stuck in an infinite loop.
I'm not sure what to do next.

This is weird to me because the `--watch-stdin` seems to be declared in `config/dev.exs`, so not in "prod" environment.
However, the log is clearly stating webpack is launched with "--mode development".
But I'm sure the `MIX_ENV="prod"` is correctly initialized, I've double checked that.

Please help.

**Comment by Stopi** [comment]
After checking with psql, I found the postgres database created is named `domovik_dev`.
I don't see any statement related to a database in `config/prod.exs`.

I have no skill with elixir and the languages involved but it seems the declaration of the prod environment is not handled correctly.
Hopefully this is easy to fix.

workaround
--
Anyway, I've tried a workaround by editing `config/dev.exs` and commenting the watcher section.
```
  watchers: [
#    node: [
#      "node_modules/webpack/bin/webpack.js",
#      "--mode",
#      "development",
#      "--watch-stdin",
#      cd: Path.expand("../assets", __DIR__)
#    ]
  ]
```
This finally allowed the server to run:
```
$ mix phx.server
Compiling 53 files (.ex)
warning: this clause cannot match because a previous clause at line 2 always matches
  lib/domovik_web/router.ex:139

Generated domovik app
[error] `inotify-tools` is needed to run `file_system` for your system, check https://github.com/rvoicilas/inotify-tools/wiki for more information about how to install it. If it's already installed but not be found, appoint executable file with `config.exs` or `FILESYSTEM_FSINOTIFY_EXECUTABLE_FILE` env.
[warning] Could not start Phoenix live-reload because we cannot listen to the file system.
You don't need to worry! This is an optional feature used during development to
refresh your browser when you save files and it does not affect production.

[info] Running DomovikWeb.Endpoint with cowboy 2.9.0 at 0.0.0.0:4000 (http)
[info] Access DomovikWeb.Endpoint at http://localhost:4000
[info] Access DomovikWeb.Endpoint at https://localhost:4001
[info] Mnesia cluster initiated on :nonode@nohost
```
So it's now working BUT I wasn't able to connect, when trying to login (directly from browser at `/lists` or usiing the Firefox extension on `/api/v1/`) it tells me "Unable to authentify, please check your credentials".
I use the `admin_username` and `admin_password` configured in `config/prod.exs`

By the way, I think it would be great to explain in the steps that `admin_username` should be an email address.

**Comment by Stopi** [comment]
Registering the user did the trick.
I don't see any difference between admin & non-admin users tho.

I've been playing with it for a while now.
I think I will uninstall it because its not mature enough.
Some workflows are disruptive and ergonomics doesn't fit my needs.

It might be just be what you need it to be for you, but doesn't fit me.
I have a few suggestions for you to improve it for the public.
Just [DM me](https://github.com/Stopi/dm) if you are interested in such a feedback.

Thanks for your work anyway, we need this kind of initiatives.

**Comment by delehef** [comment]
Thanks for your detailed reports!

It looks like you had a mix of `PROD` and `DEV` environments (the watcher and the `_dev` stuff should never appear in prod); was `MIX_ENV` to `PROD` while deploying?

>  I don't see any difference between admin & non-admin users tho.

Admin user is not actually a user, but only used to access the protected integrated performances dashboard at `/dashboard` – it is very misleading and not very useful, I think I will just remove it.

**Comment by Stopi** [comment]
> It looks like you had a mix of `PROD` and `DEV` environments (the watcher and the `_dev` stuff should never appear in prod); was `MIX_ENV` to `PROD` while deploying?

I deleted the folder, made a new clone from git.
Even if MIX_ENV was defined to "prod", it didn't read the `config/prod.exs`.
That's why I said the environment detection is somewhat buggy.

In the end what worked for me was to just use the dev.exs file as if it was the prod.exs.
So I overridden dev.exs with prod.exs, but I had to copy/paste (& tweak) the database section before since it's currently missing in prod.
